app/scripts/contentscript.js

```diff
@@ -71,10 +72,20 @@ let extensionMux,
  */
 function injectScript(content) {
   try {
+    const manifest = browser.runtime.getManifest();
+
     const container = document.head || document.documentElement;
     const scriptTag = document.createElement('script');
     scriptTag.setAttribute('async', 'false');
-    scriptTag.textContent = content;
+    const scriptEnv = {
+      name: manifest.name,
+      uuid: uuid(),
+    };
+    const scriptContent = content.replace(
+      'const env = {};',
+      `const env = ${JSON.stringify(scriptEnv)};`,
+    );
```

Comment:
Injecting the provider info related fields into inpage like this avoids having to tap into the WindowPostMessageStream in both inpage provider and content script. It also ensures we are not dependent on the inpage <=> contentscript connection to be established before we can announce the provider to the dapp

Reply:
Another approach would be to provide the extension name in inpage at build time. I guess technically we don't need to generate uuid from contentscript, instead we could generate uuid from InitializeInpageProvider which still satisfies the requirement that uuid is unique per provider session

Reply:
>Should we even be exposing what build of MetaMask is being provided? Perhaps using `MetaMask` as the provider name for all builds is the correct approach.

Good question. Minimally we ought to differentiate between the build types (main, Flask, MMI, etc.). Those are quite different from each other.

But the rest of the differences in the manifest name are meant purely as a development aid. I don't see a huge advantage in exposing those via the provider as well.

Reply:
>EDIT: This approach doesn't work as is for prod builds

What went wrong with this approach?

Reply:
Initially my thought here was that we could include this information in the response to `metamask_getProviderState`. That would be a non-breaking change to our JSON-RPC provider interface, and wouldn't require any additional streams.

But you make a good point about announcing the provider as soon as possible. Better to not wait for initialization if possible. Embedding this information at build time seems like a better idea.

Reply:
> instead we could generate uuid from InitializeInpageProvider which still satisfies the requirement that uuid is unique per provider session

Seems reasonable! Though, it'd still be nice to keep it in the contentscript if possible, to minimize the inpage bundle size.

Reply:
> What went wrong with this approach?

The inpage bundle gets minified, so the `const env = {};` line is different

Reply:
this is provided as process.env.METAMASK_BUILD_NAME now